compiler/rustc_mir_build/src/build/matches/test.rs

```diff
@@ -409,7 +409,40 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
         let deref_ty = match *ty.kind() {
             ty::Ref(_, deref_ty, _) => deref_ty,
-            _ => bug!("non_scalar_compare called on non-reference type: {}", ty),
+            _ => {
+                trace!("non_scalar_compare called on non-reference type: {}", ty);
+                // Backcompat hack: due to non-structural matches not being a hard error, we can
+                // reach this for types that have manual `Eq` or `PartialEq` impls.
+                assert!(!ty.is_structural_eq_shallow(self.hir.tcx()));
+                let ref_ty = self.hir.tcx().mk_imm_ref(self.hir.tcx().lifetimes.re_erased, ty);
+                // let y = &place;
+                let y = self.temp(ref_ty, source_info.span);
+                self.cfg.push_assign(
+                    block,
+                    source_info,
+                    y,
+                    Rvalue::Ref(self.hir.tcx().lifetimes.re_erased, BorrowKind::Shared, place),
+                );
+                val = Operand::Move(y);
+                // let temp = expect;
+                let temp = self.temp(ty, source_info.span);
+                self.cfg.push_assign(
+                    block,
+                    source_info,
+                    temp,
+                    Rvalue::Use(expect),
+                );
+                // reftemp = &temp;
+                let reftemp = self.temp(ref_ty, source_info.span);
+                self.cfg.push_assign(
+                    block,
+                    source_info,
+                    reftemp,
+                    Rvalue::Ref(self.hir.tcx().lifetimes.re_erased, BorrowKind::Shared, temp),
+                );
+                expect = Operand::Move(reftemp);
+                ty
+            },
```

Comment:
I still don't fully understand why this is needed. This is just auto-ref-ing everything that's not a reference? Seems odd, what if it's references but a mismatch of their number?

Anyway at least provide examples for both the `ty::Ref` and non-`ty::Ref` cases to understand why this is hit.

Reply:
I managed to eliminate this entirely and instead making these cases explicit in the `recur` function